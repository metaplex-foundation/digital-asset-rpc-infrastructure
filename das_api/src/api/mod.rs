use crate::DasApiError;
use async_trait::async_trait;
use digital_asset_types::rpc::filter::SearchConditionType;
use digital_asset_types::rpc::response::AssetList;
use digital_asset_types::rpc::{filter::AssetSorting, response::GetGroupingResponse};
use digital_asset_types::rpc::{Asset, AssetProof, Interface, OwnershipModel, RoyaltyModel};
use open_rpc_derive::{document_rpc, rpc};
use open_rpc_schema::schemars::JsonSchema;
use serde::{Deserialize, Serialize};

mod api_impl;
pub use api_impl::*;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAssetsByGroup {
    pub group_key: String,
    pub group_value: String,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAssetsByOwner {
    pub owner_address: String,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAsset {
    pub id: String,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]

pub struct GetAssetsByCreator {
    pub creator_address: String,
    pub only_verified: Option<bool>,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct SearchAssets {
    pub negate: Option<bool>,
    pub condition_type: Option<SearchConditionType>,
    pub interface: Option<Interface>,
    pub owner_address: Option<String>,
    pub owner_type: Option<OwnershipModel>,
    pub creator_address: Option<String>,
    pub creator_verified: Option<bool>,
    pub authority_address: Option<String>,
    pub grouping: Option<(String, String)>,
    pub delegate: Option<Vec<u8>>,
    pub frozen: Option<bool>,
    pub supply: Option<u64>,
    pub supply_mint: Option<String>,
    pub compressed: Option<bool>,
    pub compressible: Option<bool>,
    pub royalty_target_type: Option<RoyaltyModel>,
    pub royalty_target: Option<String>,
    pub royalty_amount: Option<u32>,
    pub burnt: Option<bool>,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
    #[serde(default)]
    pub json_uri: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAssetsByAuthority {
    pub authority_address: String,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetGrouping {
    pub group_key: String,
    pub group_value: String,
}

#[document_rpc]
#[async_trait]
pub trait ApiContract: Send + Sync + 'static {
    async fn check_health(&self) -> Result<(), DasApiError>;
    #[rpc(
        name = "getAssetProof",
        params = "named",
        summary = "Get a merkle proof for a compressed asset by its ID"
    )]
    async fn get_asset_proof(&self, payload: GetAsset) -> Result<AssetProof, DasApiError>;
    #[rpc(
        name = "getAsset",
        params = "named",
        summary = "Get an asset by its ID"
    )]
    async fn get_asset(&self, payload: GetAsset) -> Result<Asset, DasApiError>;
    #[rpc(
        name = "getAssetsByOwner",
        params = "named",
        summary = "Get a list of assets owned by an address"
    )]
    async fn get_assets_by_owner(
        &self,
        payload: GetAssetsByOwner,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getAssetsByGroup",
        params = "named",
        summary = "Get a list of assets by a group key and value"
    )]
    async fn get_assets_by_group(
        &self,
        payload: GetAssetsByGroup,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getAssetsByCreator",
        params = "named",
        summary = "Get a list of assets created by an address"
    )]
    async fn get_assets_by_creator(
        &self,
        payload: GetAssetsByCreator,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getAssetsByAuthority",
        params = "named",
        summary = "Get a list of assets with a specific authority"
    )]
    async fn get_assets_by_authority(
        &self,
        payload: GetAssetsByAuthority,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "searchAssets",
        params = "named",
        summary = "Search for assets by a variety of parameters"
    )]
    async fn search_assets(&self, payload: SearchAssets) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getGrouping",
        params = "named",
        summary = "Get a list of assets grouped by a specific authority"
    )]
    async fn get_grouping(&self, payload: GetGrouping) -> Result<GetGroupingResponse, DasApiError>;
}
