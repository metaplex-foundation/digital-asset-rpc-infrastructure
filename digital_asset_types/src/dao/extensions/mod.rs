pub mod asset;
pub mod asset_authority;
pub mod asset_creators;
pub mod asset_data;
pub mod asset_grouping;
pub mod asset_v1_account_attachment;
pub mod instruction;
