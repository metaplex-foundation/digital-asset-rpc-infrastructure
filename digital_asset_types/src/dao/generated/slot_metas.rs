//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.8

use sea_orm::entity::prelude::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "slot_metas")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub slot: i64,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
