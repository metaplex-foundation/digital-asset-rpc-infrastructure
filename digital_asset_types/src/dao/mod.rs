#![allow(ambiguous_glob_reexports)]
mod full_asset;
mod generated;
pub mod scopes;
use self::sea_orm_active_enums::{
    OwnerType, RoyaltyTargetType, SpecificationAssetClass, SpecificationVersions,
};
pub use full_asset::*;
pub use generated::*;
use sea_orm::{
    entity::*,
    sea_query::Expr,
    sea_query::{ConditionType, IntoCondition, SimpleExpr},
    Condition, DbErr, RelationDef,
};
use serde::{Deserialize, Serialize};

pub struct GroupingSize {
    pub size: u64,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct PageOptions {
    pub limit: u64,
    pub page: Option<u64>,
    pub before: Option<Vec<u8>>,
    pub after: Option<Vec<u8>>,
    pub cursor: Option<Cursor>,
}

#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq)]
pub struct Cursor {
    pub id: Option<Vec<u8>>,
}

pub enum Pagination {
    Keyset {
        before: Option<Vec<u8>>,
        after: Option<Vec<u8>>,
    },
    Page {
        page: u64,
    },
    Cursor(Cursor),
}

#[derive(Debug, Clone, PartialEq)]
pub struct SearchAssetsQuery {
    // Conditions
    pub negate: Option<bool>,
    /// Defaults to [ConditionType::All]
    pub condition_type: Option<ConditionType>,
    pub specification_version: Option<SpecificationVersions>,
    pub specification_asset_class: Option<SpecificationAssetClass>,
    pub owner_address: Option<Vec<u8>>,
    pub owner_type: Option<OwnerType>,
    pub creator_address: Option<Vec<u8>>,
    pub creator_verified: Option<bool>,
    pub authority_address: Option<Vec<u8>>,
    pub grouping: Option<(String, String)>,
    pub delegate: Option<Vec<u8>>,
    pub frozen: Option<bool>,
    pub supply: Option<u64>,
    pub supply_mint: Option<Vec<u8>>,
    pub compressed: Option<bool>,
    pub compressible: Option<bool>,
    pub royalty_target_type: Option<RoyaltyTargetType>,
    pub royalty_target: Option<Vec<u8>>,
    pub royalty_amount: Option<u32>,
    pub burnt: Option<bool>,
    pub json_uri: Option<String>,
    pub name: Option<Vec<u8>>,
}

impl SearchAssetsQuery {
    pub const fn count_conditions(&self) -> usize {
        // Initialize counter
        let mut num_conditions = 0;
        if self.specification_version.is_some() {
            num_conditions += 1;
        }
        if self.specification_asset_class.is_some() {
            num_conditions += 1;
        }
        if self.owner_address.is_some() {
            num_conditions += 1;
        }
        if self.owner_type.is_some() {
            num_conditions += 1;
        }
        if self.delegate.is_some() {
            num_conditions += 1;
        }
        if self.frozen.is_some() {
            num_conditions += 1;
        }
        if self.supply.is_some() {
            num_conditions += 1;
        }
        if self.supply_mint.is_some() {
            num_conditions += 1;
        }
        if self.compressed.is_some() {
            num_conditions += 1;
        }
        if self.compressible.is_some() {
            num_conditions += 1;
        }
        if self.royalty_target_type.is_some() {
            num_conditions += 1;
        }
        if self.royalty_target.is_some() {
            num_conditions += 1;
        }
        if self.royalty_amount.is_some() {
            num_conditions += 1;
        }
        if self.burnt.is_some() {
            num_conditions += 1;
        }
        if self.creator_address.is_some() {
            num_conditions += 1;
        }
        if self.creator_address.is_some() {
            num_conditions += 1;
        }
        if self.grouping.is_some() {
            num_conditions += 1;
        }
        if self.json_uri.is_some() {
            num_conditions += 1;
        }
        if self.name.is_some() {
            num_conditions += 1;
        }

        num_conditions
    }

    pub fn conditions(&self) -> Result<(Condition, Vec<RelationDef>), DbErr> {
        let mut conditions = match self.condition_type {
            // None --> default to all when no option is provided
            None | Some(ConditionType::All) => Condition::all(),
            Some(ConditionType::Any) => Condition::any(),
        };

        let mut joins = Vec::new();

        conditions = conditions
            .add_option(
                self.specification_version
                    .clone()
                    .map(|x| asset::Column::SpecificationVersion.eq(x)),
            )
            .add_option(
                self.specification_asset_class
                    .clone()
                    .map(|x| asset::Column::SpecificationAssetClass.eq(x)),
            )
            .add_option(
                self.owner_address
                    .to_owned()
                    .map(|x| asset::Column::Owner.eq(x)),
            )
            .add_option(
                self.owner_type
                    .clone()
                    .map(|x| asset::Column::OwnerType.eq(x)),
            )
            .add_option(
                self.delegate
                    .to_owned()
                    .map(|x| asset::Column::Delegate.eq(x)),
            )
            .add_option(self.frozen.map(|x| asset::Column::Frozen.eq(x)))
            .add_option(self.supply.map(|x| asset::Column::Supply.eq(x)))
            .add_option(
                self.supply_mint
                    .to_owned()
                    .map(|x| asset::Column::SupplyMint.eq(x)),
            )
            .add_option(self.compressed.map(|x| asset::Column::Compressed.eq(x)))
            .add_option(self.compressible.map(|x| asset::Column::Compressible.eq(x)))
            .add_option(
                self.royalty_target_type
                    .clone()
                    .map(|x| asset::Column::RoyaltyTargetType.eq(x)),
            )
            .add_option(
                self.royalty_target
                    .to_owned()
                    .map(|x| asset::Column::RoyaltyTarget.eq(x)),
            )
            .add_option(
                self.royalty_amount
                    .map(|x| asset::Column::RoyaltyAmount.eq(x)),
            )
            .add_option(self.burnt.map(|x| asset::Column::Burnt.eq(x)));

        if let Some(c) = self.creator_address.to_owned() {
            conditions = conditions.add(asset_creators::Column::Creator.eq(c));
        }

        // Without specifying the creators themselves, there is no index being hit.
        // So in some rare scenarios, this query could be very slow.
        if let Some(cv) = self.creator_verified.to_owned() {
            conditions = conditions.add(asset_creators::Column::Verified.eq(cv));
        }

        // If creator_address or creator_verified is set, join with asset_creators
        if self.creator_address.is_some() || self.creator_verified.is_some() {
            let rel = asset_creators::Relation::Asset
                .def()
                .rev()
                .on_condition(|left, right| {
                    Expr::tbl(right, asset_creators::Column::AssetId)
                        .eq(Expr::tbl(left, asset::Column::Id))
                        .into_condition()
                });
            joins.push(rel);
        }

        if let Some(a) = self.authority_address.to_owned() {
            conditions = conditions.add(asset_authority::Column::Authority.eq(a));
            let rel = asset_authority::Relation::Asset
                .def()
                .rev()
                .on_condition(|left, right| {
                    Expr::tbl(right, asset_authority::Column::AssetId)
                        .eq(Expr::tbl(left, asset::Column::Id))
                        .into_condition()
                });
            joins.push(rel);
        }

        if let Some(g) = self.grouping.to_owned() {
            let cond = Condition::all()
                .add(asset_grouping::Column::GroupKey.eq(g.0))
                .add(asset_grouping::Column::GroupValue.eq(g.1));
            conditions = conditions.add(cond);
            let rel = asset_grouping::Relation::Asset
                .def()
                .rev()
                .on_condition(|left, right| {
                    Expr::tbl(right, asset_grouping::Column::AssetId)
                        .eq(Expr::tbl(left, asset::Column::Id))
                        .into_condition()
                });
            joins.push(rel);
        }

        if let Some(ju) = self.json_uri.to_owned() {
            let cond = Condition::all().add(asset_data::Column::MetadataUrl.eq(ju));
            conditions = conditions.add(cond);
            let rel = asset_data::Relation::Asset
                .def()
                .rev()
                .on_condition(|left, right| {
                    Expr::tbl(right, asset_data::Column::Id)
                        .eq(Expr::tbl(left, asset::Column::AssetData))
                        .into_condition()
                });
            joins.push(rel);
        }

        if let Some(n) = self.name.to_owned() {
            let name_as_str = std::str::from_utf8(&n).map_err(|_| {
                DbErr::Custom(
                    "Could not convert raw name bytes into string for comparison".to_owned(),
                )
            })?;

            let name_expr =
                SimpleExpr::Custom(format!("chain_data->>'name' LIKE '%{}%'", name_as_str));

            conditions = conditions.add(name_expr);
            let rel = asset_data::Relation::Asset
                .def()
                .rev()
                .on_condition(|left, right| {
                    Expr::tbl(right, asset_data::Column::Id)
                        .eq(Expr::tbl(left, asset::Column::AssetData))
                        .into_condition()
                });
            joins.push(rel);
        }

        Ok((
            match self.negate {
                None | Some(false) => conditions,
                Some(true) => conditions.not(),
            },
            joins,
        ))
    }
}
