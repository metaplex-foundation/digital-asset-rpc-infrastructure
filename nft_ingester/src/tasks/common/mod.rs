use {
    super::{BgTask, FromTaskData, IngesterError, IntoTaskData, TaskData},
    async_trait::async_trait,
    chrono::{NaiveDateTime, Utc},
    digital_asset_types::dao::asset_data,
    futures::future::BoxFuture,
    log::debug,
    program_transformers::{DownloadMetadataInfo, DownloadMetadataNotifier},
    reqwest::{Client, ClientBuilder},
    sea_orm::*,
    serde::{Deserialize, Serialize},
    std::{
        fmt::{Display, Formatter},
        time::Duration,
    },
    tokio::sync::mpsc::UnboundedSender,
    url::Url,
};

pub fn create_download_metadata_notifier(
    bg_task_sender: UnboundedSender<TaskData>,
) -> DownloadMetadataNotifier {
    Box::new(
        move |info: DownloadMetadataInfo| -> BoxFuture<
            'static,
            Result<(), Box<dyn std::error::Error + Send + Sync>>,
        > {
            let (asset_data_id, uri) = info.into_inner();
            let task = DownloadMetadata {
                asset_data_id,
                uri,
                created_at: Some(Utc::now().naive_utc()),
            };
            let task = task
                .into_task_data()
                .and_then(|task| {
                    bg_task_sender.send(task).map_err(Into::into)
                })
                .map_err(Into::into);
            Box::pin(async move { task })
        },
    )
}

const TASK_NAME: &str = "DownloadMetadata";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadMetadata {
    pub asset_data_id: Vec<u8>,
    pub uri: String,
    #[serde(skip_serializing)]
    pub created_at: Option<NaiveDateTime>,
}

impl DownloadMetadata {
    pub fn sanitize(&mut self) {
        self.uri = self.uri.trim().replace('\0', "");
    }
}

impl IntoTaskData for DownloadMetadata {
    fn into_task_data(self) -> Result<TaskData, IngesterError> {
        let ts = self.created_at;
        let data =
            serde_json::to_value(self).map_err(<serde_json::Error as Into<IngesterError>>::into)?;
        Ok(TaskData {
            name: TASK_NAME,
            data,
            created_at: ts,
        })
    }
}

impl FromTaskData<DownloadMetadata> for DownloadMetadata {
    fn from_task_data(data: TaskData) -> Result<Self, IngesterError> {
        serde_json::from_value(data.data).map_err(|e| e.into())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadMetadataTask {
    pub lock_duration: Option<i64>,
    pub max_attempts: Option<i16>,
    pub timeout: Option<Duration>,
}

impl DownloadMetadataTask {
    async fn request_metadata(
        uri: String,
        timeout: Duration,
    ) -> Result<serde_json::Value, IngesterError> {
        let client = ClientBuilder::new().timeout(timeout).build()?;
        let response = Client::get(&client, uri) // Need to check for malicious sites ?
            .send()
            .await?;

        if response.status() != reqwest::StatusCode::OK {
            Err(IngesterError::HttpError {
                status_code: response.status().as_str().to_string(),
            })
        } else {
            let val: serde_json::Value = response.json().await?;
            Ok(val)
        }
    }
}

#[derive(FromQueryResult, Debug, Default, Clone, Eq, PartialEq)]
struct MetadataUrl {
    pub metadata_url: String,
}

#[async_trait]
impl BgTask for DownloadMetadataTask {
    fn name(&self) -> &'static str {
        TASK_NAME
    }

    fn lock_duration(&self) -> i64 {
        self.lock_duration.unwrap_or(5)
    }

    fn max_attempts(&self) -> i16 {
        self.max_attempts.unwrap_or(3)
    }

    async fn task(
        &self,
        db: &DatabaseConnection,
        data: serde_json::Value,
    ) -> Result<(), IngesterError> {
        let download_metadata: DownloadMetadata = serde_json::from_value(data)?;
        let meta_url = Url::parse(&download_metadata.uri);
        let body = match meta_url {
            Ok(_) => {
                DownloadMetadataTask::request_metadata(
                    download_metadata.uri.clone(),
                    self.timeout.unwrap_or(Duration::from_secs(3)),
                )
                .await?
            }
            _ => serde_json::Value::String("Invalid Uri".to_string()), //TODO -> enumize this.
        };

        let query = asset_data::Entity::find_by_id(download_metadata.asset_data_id.clone())
            .select_only()
            .column(asset_data::Column::MetadataUrl)
            .build(DbBackend::Postgres);

        match MetadataUrl::find_by_statement(query).one(db).await? {
            Some(asset) => {
                if asset.metadata_url != download_metadata.uri {
                    debug!(
                        "skipping download metadata of old URI for {:?}",
                        bs58::encode(download_metadata.asset_data_id.clone()).into_string()
                    );
                    return Ok(());
                }
            }
            None => {
                return Err(IngesterError::UnrecoverableTaskError(format!(
                    "failed to find URI in database for {:?}",
                    bs58::encode(download_metadata.asset_data_id.clone()).into_string()
                )));
            }
        }

        let model = asset_data::ActiveModel {
            id: Unchanged(download_metadata.asset_data_id.clone()),
            metadata: Set(body),
            reindex: Set(Some(false)),
            ..Default::default()
        };
        debug!(
            "download metadata for {:?}",
            bs58::encode(download_metadata.asset_data_id.clone()).into_string()
        );
        asset_data::Entity::update(model)
            .filter(asset_data::Column::Id.eq(download_metadata.asset_data_id.clone()))
            .filter(
                Condition::all()
                    .add(asset_data::Column::MetadataUrl.eq(download_metadata.uri.clone())),
            )
            .exec(db)
            .await
            .map(|_| ())
            .map_err(|db| {
                IngesterError::TaskManagerError(format!(
                    "Database error with {}, error: {}",
                    self.name(),
                    db
                ))
            })?;

        if meta_url.is_err() {
            return Err(IngesterError::UnrecoverableTaskError(format!(
                "Failed to parse URI: {}",
                download_metadata.uri
            )));
        }
        Ok(())
    }
}

impl Display for DownloadMetadata {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "DownloadMetadata from {} for {:?}",
            self.uri, self.asset_data_id
        )
    }
}
