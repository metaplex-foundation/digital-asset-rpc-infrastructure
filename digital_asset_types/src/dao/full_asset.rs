use super::{asset_v1_account_attachments, tokens};
use crate::dao::{
    asset, asset_authority, asset_creators, asset_data, asset_grouping, token_accounts,
};

#[derive(Clone, Debug, PartialEq)]
pub struct FullAssetGroup {
    pub id: i64,
    pub asset_id: Vec<u8>,
    pub group_key: String,
    pub group_value: Option<String>,
    pub seq: Option<i64>,
    pub slot_updated: Option<i64>,
    pub verified: bool,
    pub group_info_seq: Option<i64>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct FullAsset {
    pub asset: asset::Model,
    pub data: asset_data::Model,
    pub mint_with_ta: Option<(tokens::Model, Option<token_accounts::Model>)>,
    pub authorities: Vec<asset_authority::Model>,
    pub creators: Vec<asset_creators::Model>,
    pub inscription: Option<asset_v1_account_attachments::Model>,
    pub groups: Vec<(asset_grouping::Model, Option<asset_data::Model>)>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct AssetRelated {
    pub authorities: Vec<asset_authority::Model>,
    pub creators: Vec<asset_creators::Model>,
    pub groups: Vec<asset_grouping::Model>,
}

pub struct FullAssetList {
    pub list: Vec<FullAsset>,
}
