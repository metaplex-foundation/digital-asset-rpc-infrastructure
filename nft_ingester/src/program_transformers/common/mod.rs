use crate::IngesterError;
use blockbuster::programs::bubblegum::ChangeLogEventV1;
use digital_asset_types::dao::{backfill_items, cl_items};
use sea_orm::{entity::*, query::*, sea_query::OnConflict, DatabaseTransaction, DbBackend};

pub mod task;

pub async fn save_changelog_event(
    change_log_event: &ChangeLogEventV1,
    slot: u64,
    txn: &DatabaseTransaction,
) -> Result<u64, IngesterError> {
    insert_change_log(change_log_event, slot, txn, false).await?;
    Ok(change_log_event.seq)
}

fn node_idx_to_leaf_idx(index: i64, tree_height: u32) -> i64 {
    index - 2i64.pow(tree_height)
}

pub async fn insert_change_log(
    change_log_event: &ChangeLogEventV1,
    slot: u64,
    txn: &DatabaseTransaction,
    filling: bool,
) -> Result<(), IngesterError> {
    let mut i: i64 = 0;
    let depth = change_log_event.path.len() - 1;
    let tree_id = change_log_event.id.as_ref();
    for p in change_log_event.path.iter() {
        let node_idx = p.index as i64;
        println!(
            "seq {}, index {} level {}, node {:?}",
            change_log_event.seq,
            p.index,
            i,
            bs58::encode(p.node).into_string()
        );
        let leaf_idx = if i == 0 {
            Some(node_idx_to_leaf_idx(node_idx, depth as u32))
        } else {
            None
        };

        let item = cl_items::ActiveModel {
            tree: Set(tree_id.to_vec()),
            level: Set(i),
            node_idx: Set(node_idx),
            hash: Set(p.node.as_ref().to_vec()),
            seq: Set(change_log_event.seq as i64),
            leaf_idx: Set(leaf_idx),
            ..Default::default()
        };
        i += 1;
        let mut query = cl_items::Entity::insert(item)
            .on_conflict(
                OnConflict::columns([cl_items::Column::Tree, cl_items::Column::NodeIdx])
                    .update_columns([cl_items::Column::Hash, cl_items::Column::Seq])
                    .to_owned(),
            )
            .build(DbBackend::Postgres);
        if !filling {
            query.sql = format!("{} WHERE excluded.seq > cl_items.seq", query.sql);
        }
        txn.execute(query)
            .await
            .map_err(|db_err| IngesterError::StorageWriteError(db_err.to_string()))?;
    }

    // If and only if the entire path of nodes was inserted into the `cl_items` table, then insert
    // a single row into the `backfill_items` table.  This way if an incomplete path was inserted
    // into `cl_items` due to an error, a gap will be created for the tree and the backfiller will
    // fix it.
    if i - 1 == depth as i64 {
        // See if the tree already exists in the `backfill_items` table.
        let rows = backfill_items::Entity::find()
            .filter(backfill_items::Column::Tree.eq(tree_id))
            .limit(1)
            .all(txn)
            .await?;

        // If the tree does not exist in `backfill_items` and the sequence number is greater than 1,
        // then we know we will need to backfill the tree from sequence number 1 up to the current
        // sequence number.  So in this case we set at flag to force checking the tree.
        let force_chk = rows.is_empty() && change_log_event.seq > 1;

        println!("Adding to backfill_items table at level {}", i - 1);
        let item = backfill_items::ActiveModel {
            tree: Set(tree_id.to_vec()),
            seq: Set(change_log_event.seq as i64),
            slot: Set(slot as i64),
            force_chk: Set(force_chk),
            backfilled: Set(false),
            failed: Set(false),
            ..Default::default()
        };

        backfill_items::Entity::insert(item).exec(txn).await?;
    }

    Ok(())
    //TODO -> set maximum size of path and break into multiple statements
}
