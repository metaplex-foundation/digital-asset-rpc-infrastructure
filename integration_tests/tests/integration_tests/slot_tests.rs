use super::common::*;
use das_api::api::ApiContract;
use digital_asset_types::dao::slot_metas;
use function_name::named;
use sea_orm::{ActiveValue::Set, EntityTrait};
use serial_test::serial;

#[tokio::test]
#[serial]
#[named]
async fn test_get_slot() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new(name.clone()).await;

    apply_migrations_and_delete_data(setup.db.clone()).await;
    let model = slot_metas::ActiveModel { slot: Set(12345) };
    slot_metas::Entity::insert(model)
        .exec(&setup.das_api.get_connection())
        .await
        .unwrap();

    let response = setup.das_api.get_slot(None).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}
