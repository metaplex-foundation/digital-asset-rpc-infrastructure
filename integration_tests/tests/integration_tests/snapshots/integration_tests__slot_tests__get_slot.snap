---
source: integration_tests/tests/integration_tests/slot_tests.rs
expression: response
---
12345
